corehq/sql_db/util.py

```diff
@@ -150,3 +153,53 @@ def _get_all_nested_subclasses(cls):
         if subclass not in seen:
             seen.add(subclass)
             yield subclass
+
+
+@memoized
+def get_standby_databases():
+    standby_dbs = []
+    for db_alias in settings.DATABASES:
+        with db.connections[db_alias].cursor() as cursor:
+            cursor.execute("SELECT pg_is_in_recovery()")
+            [(is_standby, )] = cursor.fetchall()
+            if is_standby:
+                standby_dbs.append(db_alias)
+    return standby_dbs
+
+
+STANDBY_DELAY_FRESHNESS = 30
+
+
+@quickcache(['db_alias'], timeout=STANDBY_DELAY_FRESHNESS)
+def get_replication_delay_for_standby(db_alias):
+    """
+    Finds the replication delay for given database by running a SQL query on standby database.
+        See https://www.postgresql.org/message-id/CADKbJJWz9M0swPT3oqe8f9+tfD4-F54uE6Xtkh4nERpVsQnjnw@mail.gmail.com
+
+    If the given database is not a standby database, zero delay is returned
+    If standby process (wal_receiver) is not running on standby a `VERY_LARGE_DELAY` is returned
+    """
+    if db_alias not in get_standby_databases():
+        return 0
+    # used to indicate that the wal_receiver process on standby is not running
+    VERY_LARGE_DELAY = 100000
+    sql = """
+    SELECT
+    CASE
+        WHEN NOT EXISTS (SELECT 1 FROM pg_stat_wal_receiver) THEN {delay}
+        WHEN pg_last_xlog_receive_location() = pg_last_xlog_replay_location() THEN 0
+        ELSE EXTRACT (EPOCH FROM now() - pg_last_xact_replay_timestamp())::INTEGER
+    END
+    AS replication_lag;
+    """.format(delay=VERY_LARGE_DELAY)
+    with db.connections[db_alias].cursor() as cursor:
+        cursor.execute(sql)
+        [(delay, )] = cursor.fetchall()
+        return delay
+
+def filter_out_stale_standbys(dbs, delay_threshold):
+    return [
+        db
+        for db in dbs
+        if get_replication_delay_for_standby(db) <= delay_threshold
```

Comment:
could pre-filter databases that we know aren't standbys to avoid hitting the cache for them

Reply:
That's done in `get_replication_delay_for_standby `, it returns zero if it's not a standby db.